taichi/backends/codegen_opengl.cpp

```diff
@@ -0,0 +1,534 @@
+      irpass::print(ir);
+    }
+  }
+
+  irpass::lower(ir);
+  irpass::re_id(ir);
+  if (print_ir) {
+    TI_TRACE("Lowered:");
+    irpass::print(ir);
+  }
+
+  irpass::typecheck(ir);
+  irpass::re_id(ir);
+  if (print_ir) {
+    TI_TRACE("Typechecked:");
+    irpass::print(ir);
+  }
+
+  irpass::demote_dense_struct_fors(ir);
+  irpass::typecheck(ir);
+  if (print_ir) {
+    TI_TRACE("Dense Struct-for demoted:");
+    irpass::print(ir);
+  }
+
+  irpass::constant_fold(ir);
+  if (prog_->config.simplify_before_lower_access) {
+    irpass::simplify(ir);
+    irpass::re_id(ir);
+    if (print_ir) {
+      TI_TRACE("Simplified I:");
+      irpass::print(ir);
+    }
+  }
+
+  if (kernel_->grad) {
+    irpass::demote_atomics(ir);
+    irpass::full_simplify(ir);
+    irpass::typecheck(ir);
+    if (print_ir) {
+      TI_TRACE("Before make_adjoint:");
+      irpass::print(ir);
+    }
+    irpass::make_adjoint(ir);
+    if (print_ir) {
+      TI_TRACE("After make_adjoint:");
+      irpass::print(ir);
+    }
+    irpass::typecheck(ir);
+  }
+
+  irpass::lower_access(ir, prog_->config.use_llvm);
+  irpass::re_id(ir);
+  if (print_ir) {
+    TI_TRACE("Access Lowered:");
+    irpass::print(ir);
+  }
+
+  irpass::die(ir);
+  irpass::re_id(ir);
+  if (print_ir) {
+    TI_TRACE("DIEd:");
+    irpass::print(ir);
+  }
+
+  irpass::flag_access(ir);
+  irpass::re_id(ir);
+  if (print_ir) {
+    TI_TRACE("Access Flagged:");
+    irpass::print(ir);
+  }
+
+  irpass::constant_fold(ir);
+  if (print_ir) {
+    TI_TRACE("Constant folded:");
+    irpass::re_id(ir);
+    irpass::print(ir);
+  }
+
+  global_tmps_buffer_size_ =
+      std::max(irpass::offload(ir).total_size, (size_t)(1));
+  if (print_ir) {
+    TI_TRACE("Offloaded:");
+    irpass::re_id(ir);
+    irpass::print(ir);
+  }
+
+  irpass::full_simplify(ir);
+  if (print_ir) {
+    TI_TRACE("Simplified II:");
+    irpass::re_id(ir);
+    irpass::print(ir);
+  }
+
+  irpass::demote_atomics(ir);
+  if (print_ir) {
+    TI_TRACE("Atomics demoted:");
+    irpass::re_id(ir);
+    irpass::print(ir);
+  }
+}
+
+void load_data(Context &ctx, void *data)
+{
+  int *data_ = (int *)data;
+  for (int i = 0; i < taichi_max_num_args; i++) {
+    int value = ctx.get_arg<int>(i);
+    data_[i] = value;
+  }
+}
+
+void save_data(Context &ctx, void *data)
+{
+  int *data_ = (int *)data;
+  for (int i = 0; i < taichi_max_num_args; i++) {
+    int value = data_[i];
+    ctx.set_arg<int>(i, value);
+  }
+}
+
+FunctionType OpenglCodeGen::gen(void)
+{
+  KernelGen codegen(kernel_, kernel_name_);
+  codegen.run(*prog_->snode_root);
+  const std::string kernel_source_code = codegen.kernel_source_code();
+  TI_INFO("\n{}", kernel_source_code);
+
+  return [kernel_source_code](Context &ctx) {
+    void *data, *data_r;
+    size_t data_size = 1024; // ...
+    data = malloc(data_size);
+    load_data(ctx, data);
+    data_r = launch_glsl_kernel(kernel_source_code, data, data_size);
+    free(data);
+    save_data(ctx, data_r);
+  };
```

Comment:
Here, in load_data, read globals & arguments from `ctx` to `data`, vise versa.

Reply:
Looks good! The `malloc`/`free` pair here seems s bit dangerous. Maybe we should use a `std::vector<uint8_t> data(data_size)`?

Reply:
Sloved using SSBO class, auto-free on destroy.